gson/src/main/java/com/google/gson/TypeAdapter.java

```diff
@@ -289,24 +289,30 @@ public final T fromJsonTree(JsonElement jsonTree) {
    * Note that we didn't need to check for nulls in our type adapter after we used nullSafe.
    */
   public final TypeAdapter<T> nullSafe() {
-    return new TypeAdapter<T>() {
-      @Override
-      public void write(JsonWriter out, T value) throws IOException {
-        if (value == null) {
-          out.nullValue();
-        } else {
-          TypeAdapter.this.write(out, value);
-        }
+    if (this.getClass() != NullSafeTypeAdapter.class) {
+      return new NullSafeTypeAdapter();
+    }
+    return this;
+  }
+
+  private final class NullSafeTypeAdapter extends TypeAdapter<T> {
```

Comment:
* I prefer static nested classes, but I also guess that it might be an exception from the preferences, since `NullSafeTypeAdapter` captures `this` anyway.
* I'm not sure if it's worth adding `toString` here since the base class doesn't (?) override `Object.toString` that would be transparent anyway.